gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1298 @@
+                detected_num_models = sum([ensemble.num_models for ensemble in target])
+
+            # 3. list of gensim models
+            elif isinstance(target[0], basemodel.BaseTopicModel):
+                ttda = np.concatenate([model.get_topics() for model in target], axis=0)
+                detected_num_models = len(target)
+
+            # unknown
+            else:
+                raise ValueError("target is of unknown type or a list of unknown types: {}".format(type(target[0])))
+
+            # new models were added, increase num_models
+            # if the user didn't provide a custon numer to use
+            if num_new_models is None:
+                self.num_models += detected_num_models
+            else:
+                self.num_models += num_new_models
+
+        else:  # memory unfriendly ensembles
+
+            ttda = []
+
+            # 1. ttda array
+            if isinstance(target.dtype.type(), (np.number, float)):
+                raise ValueError('ttda arrays cannot be added to ensembles, for which memory_friendly_ttda=False, '
+                                 'you can call convert_to_memory_friendly, but it will discard the stored gensim'
+                                 'models and only keep the relevant topic term distributions from them.')
+
+            # 2. list of ensembles
+            elif isinstance(target[0], type(self)):
+                for ensemble in target:
+                    self.tms += ensemble.tms
+                ttda = np.concatenate([ensemble.ttda for ensemble in target], axis=0)
+
+            # 3. list of gensim models
+            elif isinstance(target[0], basemodel.BaseTopicModel):
+                self.tms += target.tolist()
+                ttda = np.concatenate([model.get_topics() for model in target], axis=0)
+
+            # unknown
+            else:
+                raise ValueError("target is of unknown type or a list of unknown types: {}".format(type(target[0])))
+
+            # in this case, len(self.tms) should
+            # always match self.num_models
+            if num_new_models is not None and num_new_models + self.num_models != len(self.tms):
+                logger.info('num_new_models will be ignored. num_models should match the number of '
+                            'stored models for a memory unfriendly ensemble')
+            self.num_models = len(self.tms)
+
+        logger.info("ensemble contains {} models and {} topics now".format(self.num_models, len(self.ttda)))
+
+        if self.ttda.shape[1] != ttda.shape[1]:
+            raise ValueError(("target ttda dimensions do not match. Topics must be {} but was"
+                                "{} elements large").format(self.ttda.shape[-1], ttda.shape[-1]))
+        self.ttda = np.append(self.ttda, ttda, axis=0)
+
+        # tell recluster that the distance matrix needs to be regenerated
+        self.asymmetric_distance_matrix_outdated = True
+
+    def save(self, fname):
```

Comment:
Inherit from `utils.SaveLoad`, to get these `save` and `load` methods for free. This "standard" Gensim serialization supports additional features, such as serializing large arrays (where plain pickle will fail).

Reply:
It fails here when val is a class instead of an object:
https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/utils.py#L618

Since `hasattr(LdaModel, '_save_specials')` is True. The first argument to _save_specials will then be 'self', hence subname is missing as a positional argument.

```
TypeError: _save_specials() missing 1 required positional argument: 'subname'
```

I think it would work if it said something among
```
if hasattr(val, '_save_specials') and type(val) != type:
```
in line 609, do you see any downside to this? I'm going to commit it and see if gensims tests pass in the meantime. Or with isinstance, but apparently that causes other issues (see comment of line 609).

We need a class as one of the attributes in ensembleLda, because we need to store which models are going to be trained as children.

Reply:
I don't follow. Can you show the full traceback?

Reply:
`elda.__dict__['topic_model_class']` is `gensim.models.ldamodel.LdaModel` or other classes like LdaMulticore

```
TypeError                          Traceback (most recent call last)
~/Code/ensemblelda/gensim/gensim/utils.py in save(self, fname_or_handle, separately, sep_limit, ignore, pickle_protocol)
    696         try:
--> 697             _pickle.dump(self, fname_or_handle, protocol=pickle_protocol)
    698             logger.info("saved %s object", self.__class__.__name__)

TypeError: file must have a 'write' attribute

During handling of the above exception, another exception occurred:

TypeError                          Traceback (most recent call last)
<ipython-input-9-35ccc388ca18> in <module>
----> 1 elda.save("./bla")

~/Code/ensemblelda/gensim/gensim/utils.py in save(self, fname_or_handle, separately, sep_limit, ignore, pickle_protocol)
    698             logger.info("saved %s object", self.__class__.__name__)
    699         except TypeError:  # `fname_or_handle` does not have write attribute
--> 700             self._smart_save(fname_or_handle, separately, sep_limit, ignore, pickle_protocol=pickle_protocol)
    701 
    702 

~/Code/ensemblelda/gensim/gensim/utils.py in _smart_save(self, fname, separately, sep_limit, ignore, pickle_protocol)
    548 
    549         restores = self._save_specials(fname, separately, sep_limit, ignore, pickle_protocol,
--> 550                                        compress, subname)
    551         try:
    552             pickle(self, fname, protocol=pickle_protocol)

~/Code/ensemblelda/gensim/gensim/utils.py in _save_specials(self, fname, separately, sep_limit, ignore, pickle_protocol, compress, subname)
    612                 cfname = '.'.join((fname, attrib))
    613                 print(val, subname)
--> 614                 restores.extend(val._save_specials(cfname, None, sep_limit, ignore, pickle_protocol, compress, subname))
    615 
    616         try:

TypeError: _save_specials() missing 1 required positional argument: 'subname'
```

console output right before the line that fails:

`<class 'gensim.models.ldamodel.LdaModel'> <function SaveLoad._adapt_by_suffix.<locals>.<lambda> at 0x7f4dc8d40290>`

Reply:
the suggested change of `if hasattr(val, '_save_specials') and type(val) != type:` passes the CI tests

Reply:
Thanks @sezanzeb. I'll look at this more closely, the existing Gensim code looks wild. CC @mpenkov.

Reply:
OK, I think I see what you mean. Some object attribute is a class, and should not be saved recursively.

The `isinstance(val, SaveLoad)` mentioned in the comment looks better. That comment and code change was introduced [here](https://github.com/RaRe-Technologies/gensim/commit/9cac3843481688f79e3904e893f49163c6328789) by @gojomo, but looks like a regression to me.

@gojomo what are these "IPython reloads" that prompted this? Do we really need this?

Reply:
OK, I think I deciphered that too :) If a user explicitly reloads the `utils` module, for example from an interactive Python prompt, `SaveLoad` will be a different object in the reloaded module vs the previously loaded modules.

And then the check `isinstance(obj, SaveLoad)` may return false, even though `obj` inherits from `SaveLoad` (but it's the other, non-reloaded `SaveLoad`).

I don't think this use-case is really worth it though. @mpenkov @gojomo thoughts? I'm +1 on returning back to plain `isinstance(obj, SaveLoad)`… assuming that solves this issue @sezanzeb identified here.

Reply:
Yes, the problem is that in a dynamic environment like IPython/Jupyter, where classes may be reloaded, some older objects' reference to `SaveLoad` may not be the identity-same class object – so it's more robust to test for the capability, using Python-style 'duck-typing', than the exact class/inheritance.  As that change has been working fine for 4.5 years, & also strikes me as generally more 'hygenic' & flexible for the common case of Jupyter-notebook work, a reversion seems misguided. I'm not sure why now, for the first time in years, this method would be called without a 'self' instance, so I'd tend to think whatever new usage is calling it in a new way should change. Or, if it needs a check against being inappropriately called on classes, adding the extra check @sezanzeb proposed makes more sense than re-introducing an older fragility.

Reply:
That proposed check is not 100% though, because apart from "being a class", the object may also be a function or a module (among other things).

I can see how the reloading scenario can be useful. Although I'd expect you'll encounter other issues too if you do such runtime shenanigans, with different versions of the same module / class across your environment. I'm surprised hacking one `if` inside one method was enough.

More "hygenic" would be to go and check what we **really** want to be checking: whether the attribute is an object inheriting from SaveLoad, so we can save it separately.

Not to derail this PR, @sezanzeb did a great job. I'm mildly in favour of the more straightforward & correct `instance(obj, SaveLoad)`, but no strong preference. @mpenkov WDYT?

Reply:
@piskvorky We already have an issue to author a tutorial for better discovery (https://github.com/RaRe-Technologies/gensim/issues/2673).

For the way forward, I suggest we go with what's already in this branch, and kick off a separate issue to deal with doing serialization "the proper way", whatever we decide that to be in the future.

Reply:
The proper way is to use `SaveLoad`, there's no question about that. This discussion is on a subtle technical point: how to best test whether an object attribute is SaveLoad-serializable itself, recursively.

Reply:
for the moment maybe we can stick to `if hasattr(val, '_save_specials') and type(val) != type:` which makes things work right now without side effects. Later we can still change it to something more sophisticated.

Reply:
@sezanzeb does `isinstance(obj, SaveLoad)` solve your issue?

Reply:
I still don't see `isinstance(obj, SaveLoad)` as `more straightforward & correct'. It's more fragile in a real usage scenario, Jupyter notebooks,  that's important for developers. 

I'm not sure why this new code is, for the 1st time, now passing actual class objects into this test. That no prior usage attempted this suggests to me it's a violation of the implicit contract of `save()` – save should be serializing instances, **not** full classes that might be recursively mentioned. 

So my hunch is the best fix would be to change the new code to avoid demanding very-different things from older code (like handling full-on class objects in `_smart_save()` special serialization). Maybe the `.topic_model_class` should be a specific string name, from a controlled-vocabulary of allowable options. Or, a fully-qualified class name string, to allow reference to any loadable class, but without having an actual class-object there. (Allowing actual class-objects in what you're trying to serialize is like having lambdas or functions or user-package callback objects – a major headache for later deserialization.) The explicit documentation of `SaveLoad` could be expanded to explain that properties with references to {classes, lambdas, functions, user-package code objects} are either unsupported or strongly not-recommended.

If it's the intent for `SaveLoad` to support saving references to class objects – classes which might know how to `_save_specials()` their instances, but not the class object – then `SaveLoad` could be upgraded to handle that case, and enforce the (very logical) idea that it is only instance-objects, and not class-objects, that get to specify their own custom serialization via `_save_specials()`. One way to do that is @sezanzeb's suggested fix, to tighten the test of an object's eligilbility for `_save_specials()`. (It might also be possible to check for different handling of class-objects earlier.)

Reply:
I will try if I can just pickle the name of the class as string and load it with https://docs.python.org/3/library/importlib.html#importlib.import_module afterwards.

---

```
import pickle
import gensim
with open('dump', 'wb') as f:
    pickle.dump(gensim.models.LdaModel, f)
```
```
!cat dump
```
```
�cgensim.models.ldamodel
LdaModel
q.
```

I think this just stores the name or something. Or what do you mean with "full classes"?

Restricting it on known topic model classes by using strings in the pickling would mean that new kinds of topic models cannot be clustered with eLDA.

---

I don't think I want to break stuff in jupyter notebooks by using isinstance

Reply:
It won't likely break any polished notebooks, that are designed to run right through. Rather, when doing related development, and interactively revising code & re-running cells with live instances, the `isinstance()` check will be more fragile. So it's not a big thing for average users, but it is something I've run into when extending/fixing gensim code. 

When supplying 'full classes' (class/type objects), yes, the serialization still just writes barely more than the class name to disk. But the behavior/expectations are different – it's "as if" a richer/typed object was written, and the pickle-machinery will try to restore that object, and fail if at load-time a class of the same name is unavailable, with a possibly-unhelpful error message & no load of the rest of the state (which might help debug the situation). 

If instead storing the fully-qualified class-name as string instead, then saving/loading should reliably succeed. Then, it's potentially later code, outside the deserialization routines, that changes that string into an instantiatable class when needed. If that step fails, it can perhaps give a more helpful error message, & allow a focused retry of just that step after the user has installed/imported the named class(es).

Reply:
https://github.com/RaRe-Technologies/gensim/blob/c559de6cd3366f51af2110758f578ac72d4ec768/gensim/models/ensemblelda.py#L271-L293

Reply:
This fits the gist of what I'd suggested, and so to the extent it avoids persisting a code object & passes tests, that's good. But I believe it could be even more robust & minimal if the class reference is *never* stored as a property, but rather *always* created on demand, at the moments that it is needed, from a fully-qualified string. Then, there's no need for a custom `save()`/`load()` at all. If there is a problem at the moment the class reference is created-on-demand, the exception naturally raised there may be enough to debug the problem – or it could be caught and described with an extra error message.

Reply:
@gojomo the custom load is now obsolete.

But I don't think I can make the save function obsolete as well. I could import topic_model_class every time I query the attribute, but I think this fails when topic_model_class is not a module that can be imported from a file. For example some custom class in a jupyter notebook.